third_party/nvidia/lib/Dialect/NVWS/Transforms/LowerWarpGroup.cpp

```diff
@@ -102,31 +100,78 @@ class LowerWarpGroup : public OpRewritePattern<WarpGroupOp> {
       mlir::getUsedValuesDefinedAbove(*partition, captures);
 
     SmallVector<Value> inputs;
-    SmallVector<Value> constants;
-    for (Value capture : captures) {
-      if (!isa<BlockArgument>(capture) &&
-          isa<arith::ConstantOp>(capture.getDefiningOp())) {
-        constants.push_back(capture);
+    SmallVector<IRMapping> mappings(partitions.size());
+    SmallVector<OpBuilder> builders;
+    for (auto region : partitions) {
+      builders.push_back(OpBuilder::atBlockBegin(&region->front()));
+    }
+
+    SetVector<Operation *> opsToClone;
+    std::queue<Value> que;
+    for (auto capture : captures) {
+      que.push(capture);
+    }
+
+    while (!que.empty()) {
+      Value capture = que.front();
+      Operation *defOp = capture.getDefiningOp();
+      if (!isa<BlockArgument>(capture) && defOp && isPure(defOp) &&
+          (defOp->hasTrait<OpTrait::ConstantLike>() ||
+           isa<RankedTensorType>(capture.getType()))) {
+        for (auto operand : defOp->getOperands()) {
+          que.push(operand);
+        }
+        opsToClone.insert(defOp);
+      } else if (auto tensorTy =
+                     dyn_cast<RankedTensorType>(capture.getType())) {
+        SharedEncodingTrait sharedEnc = getSharedEncoding(tensorTy);
+        auto memdescTy = MemDescType::get(
+            tensorTy.getShape(), tensorTy.getElementType(), sharedEnc,
+            SharedMemorySpaceAttr::get(tensorTy.getContext()));
+        auto alloc = rewriter.create<LocalAllocOp>(loc, memdescTy, capture);
+        for (auto [i, region] : llvm::enumerate(partitions)) {
+          Value value = builders[i].create<LocalLoadOp>(capture.getLoc(),
+                                                        tensorTy, alloc);
+          replaceAllUsesInRegionWith(capture, value, *region);
+          mappings[i].map(capture, value);
+        }
+        inputs.push_back(alloc);
       } else {
         inputs.push_back(capture);
       }
+      que.pop();
     }
 
-    auto wsOp = rewriter.create<WarpSpecializeOp>(loc, TypeRange(), inputs);
+    opsToClone = topologicalSort(opsToClone);
+
+    for (auto [region, b, mapping] :
+         llvm::zip(partitions, builders, mappings)) {
+      for (Operation *op : opsToClone) {
+        auto copy = b.clone(*op, mapping)->getResult(0);
+        mapping.map(op->getResult(0), copy);
+        replaceAllUsesInRegionWith(op->getResult(0), copy, *region);
```

Comment:
what happens if the op has more than 1 result?

Reply:
This code was taken from the existing code in https://github.com/triton-lang/triton/blob/main/lib/Dialect/TritonGPU/Transforms/WarpSpecialization/PartitionLoops.cpp#L239-L242

We clone only constant or pure tensor ops here, that's why I think it has been working fine.

Reply:
Aha, right. Well if it was broken before then it's not a big deal